File: old/python/app_basic.py
#!/usr/bin/python3

#
# Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from functools import reduce
import numpy as np
import tensorrt
import pycuda.driver as cuda
import pycuda.autoinit
from trt_lite import TrtLite

np.set_printoptions(threshold=np.inf)

def build_engine_static(builder, input_shape):
    network = builder.create_network()
    data = network.add_input("data", tensorrt.DataType.FLOAT, input_shape)

    w = np.asarray(
        [0, 0, 0,
         0, 1, 0,
         0, 0, 0],
        dtype = np.float32)
    b = np.zeros((1,), np.float32)
    conv = network.add_convolution(data, 1, (3, 3), w, b)
    conv.stride = (1, 1)
    conv.padding = (1, 1)
    print('conv', conv.get_output(0).shape)

    network.mark_output(conv.get_output(0))
    builder.max_batch_size = 64
    return builder.build_engine(network, builder.create_builder_config())

def run_engine_static(save_and_load=False):
    batch_size = 1
    input_shape = (batch_size, 1, 5, 5)
    n = reduce(lambda x, y: x * y, input_shape)
    input_data = np.asarray(range(n), dtype=np.float32).reshape(input_shape)
    output_data = np.zeros(input_shape, dtype=np.float32)
    
    trt = TrtLite(build_engine_static, (input_shape[1:],))
    if save_and_load:
        trt.save_to_file("out.trt")
        trt = TrtLite(engine_file_path="out.trt")
    trt.print_info()

    d_buffers = trt.allocate_io_buffers(batch_size, True)

    cuda.memcpy_htod(d_buffers[0], input_data)
    trt.execute(d_buffers, batch_size)
    cuda.memcpy_dtoh(output_data, d_buffers[1])
    
    print(output_data)

def build_engine_dynamic(builder):
    network = builder.create_network(1)
    data = network.add_input("data", tensorrt.DataType.FLOAT, (-1, 1, -1, -1))

    w = np.asarray(
        [0, 0, 0,
         0, 1, 0,
         0, 0, 0],
        dtype = np.float32)
    b = np.zeros((1,), np.float32)
    conv = network.add_convolution(data, 1, (3, 3), w, b)
    conv.stride = (1, 1)
    conv.padding = (1, 1)
    print('conv', conv.get_output(0).shape)

    network.mark_output(conv.get_output(0))

    op = builder.create_optimization_profile()
    op.set_shape('data', (1, 1, 3, 3), (1, 1, 5, 5), (16, 1, 128, 128))
    config = builder.create_builder_config()
    config.add_optimization_profile(op)
    
    return builder.build_engine(network, config)

def run_engine_dynamic(save_and_load=False):
    input_shape = (1, 1, 5, 5)
    n = reduce(lambda x, y: x * y, input_shape)
    input_data = np.asarray(range(n), dtype=np.float32).reshape(input_shape)
    output_data = np.zeros(input_shape, dtype=np.float32)
    
    trt = TrtLite(build_engine_dynamic)
    if save_and_load:
        trt.save_to_file("out.trt")
        trt = TrtLite(engine_file_path="out.trt")
    trt.print_info()

    i2shape = {0: input_shape}
    d_buffers = trt.allocate_io_buffers(i2shape, True)

    cuda.memcpy_htod(d_buffers[0], input_data)
    trt.execute(d_buffers, i2shape)
    cuda.memcpy_dtoh(output_data, d_buffers[1])
    
    print(output_data)


if __name__ == '__main__':
    run_engine_static()
    run_engine_static(True)
    run_engine_dynamic()
    run_engine_dynamic(True)

File: old/samples/AppBasic.cpp
/*
 * Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "TrtLite.h"
#include "Utils.h"

simplelogger::Logger *logger = simplelogger::LoggerFactory::CreateConsoleLogger(simplelogger::TRACE);

IHostMemory *BuildNetworkProc(IBuilder *builder, void *pData) {
    BuildEngineParam *pParam = (BuildEngineParam *)pData;
    unique_ptr<INetworkDefinition> network(builder->createNetworkV2(0));
    ITensor *tensor = network->addInput("input0", DataType::kFLOAT, Dims3{pParam->nChannel, pParam->nHeight, pParam->nWidth});
    float kernel[] = {
        0.0f, 0.0f, 0.0f,
        0.0f, 1.0f, 0.0f,
        0.0f, 0.0f, 0.0f,
    };
    float bias[] = {
        0.0f,
    };
    for (int i = 0; i < 2; i++) {
        IConvolutionLayer *conv = network->addConvolutionNd(*tensor, pParam->nChannel, DimsHW(3, 3), 
            Weights{DataType::kFLOAT, kernel, sizeof(kernel) / sizeof(kernel[0])}, 
            Weights{DataType::kFLOAT, bias, sizeof(bias) / sizeof(bias[0])}
        );
        conv->setPaddingMode(PaddingMode::kSAME_LOWER);
        tensor = conv->getOutput(0);
    }
    network->markOutput(*tensor);

    unique_ptr<IBuilderConfig> config(builder->createBuilderConfig());
    if (pParam->bFp16) {
        config->setFlag(BuilderFlag::kFP16);
    }
    if (pParam->bRefit) {
        config->setFlag(BuilderFlag::kREFIT);
    }
    builder->setMaxBatchSize(pParam->nMaxBatchSize);

    return builder->buildSerializedNetwork(*network, *config);
}

int main(int argc, char** argv) {
    int iDevice = 0;
    if (argc >= 2) iDevice = atoi(argv[1]);
    ck(cudaSetDevice(iDevice));
    cudaDeviceProp prop = {};
    ck(cudaGetDeviceProperties(&prop, iDevice));
    cout << "Using " << prop.name << endl;

    BuildEngineParam param = {16, 1, 2, 8};
    param.bFp16 = true;
    param.bRefit = true;
    int nBatch = 2;

    auto trt = unique_ptr<TrtLite>(TrtLiteCreator::Create(BuildNetworkProc, &param));
    trt->PrintInfo();
    
    vector<void *> vpBuf, vdpBuf;
    vector<IOInfo> vInfo = trt->ConfigIO(nBatch);
    for (auto info : vInfo) {
        cout << info.to_string() << endl;
        
        void *pBuf = nullptr;
        pBuf = new uint8_t[info.GetNumBytes()];
        vpBuf.push_back(pBuf);

        void *dpBuf = nullptr;
        ck(cudaMalloc(&dpBuf, info.GetNumBytes()));
        vdpBuf.push_back(dpBuf);

        if (info.bInput) {
            fill((float *)pBuf, info.GetNumBytes() / sizeof(float), 1.0f);
            ck(cudaMemcpy(dpBuf, pBuf, info.GetNumBytes(), cudaMemcpyHostToDevice));
        }
    }
    trt->Execute(nBatch, vdpBuf);
    for (int i = 0; i < vInfo.size(); i++) {
        auto &info = vInfo[i];
        if (info.bInput) {
            continue;
        }
        ck(cudaMemcpy(vpBuf[i], vdpBuf[i], info.GetNumBytes(), cudaMemcpyDeviceToHost));
    }

    print((float *)vpBuf[1], nBatch * param.nChannel * param.nHeight, param.nWidth);    
    const char *szPath = "out.trt";
    trt->Save(szPath);
    
    for (int i = 0; i < vInfo.size(); i++) {
        delete[] (uint8_t *)vpBuf[i];
        ck(cudaFree(vdpBuf[i]));
    }
    
    return 0;
}
